### ISSUE TITLE
ExtensionPoint Undefined Behavior

### ISSUE THREAD

**Issue by astrelsky** [NationalSecurityAgency/ghidra#1260]
**Describe the ~~bug~~ oversight**
I'm not entirely certain if this is truly undefined behavior, it just seems the best way to describe it. This is similar to the question being asked in #1251 except I have noticed that some Extension Points won't be used every time they should be. This is because the ClassSearcher finds all classes implementing the ExtensionPoint and returns them in an unsorted set in both getClasses and getInstances methods. Many classes which make use of extension points iterate through the results until they find an instance which meets their needs. However, this is under the assumption that **only one** instance will meet their needs. Since the result from ClassSearcher is unsorted the user of an ExtensionPoint will sometimes decide to use a new implementation and will sometimes use ghidra's default implementation.

**To Reproduce**
Steps to reproduce the behavior using Demangler as an example:
1. Create a new class which implements Demangler ans satisfies its requirements.
2. Set canDemangle to always return true.
3. Set a breakpoint in GnuDemangler.demangle and a breakpoint in the new implementations demangle method.
3. Load an elf binary into ghidra and run the demangler analyzer multiple times.
4. Notice the debugger will sometimes break in GnuDemangler.demangle and sometimes in the new implementations demangle method.

**Expected behavior**
The same ExtensionPoint should be used under set conditions every time.

**Additional context**
Some ExtensionPoints, such as AbstractLibrarySupportLoader, have methods like getTier and getTierPriority and sort the results from the ClassSearcher to circumvent this. I'm not sure what the best approach would be to solve this oversight, but maybe having ExtensionPoint require such a method and having the ClassSearcher sort the result before returning would be helpful.

**Comment by ryanmkurtz** [comment]
We are thinking:

1. Sort the classpath that both the class loader and class searcher use according to some priority scheme (something like group frameworks, then features, then processors, then extensions, and subsort each group by module name)
2. Add an annotation to the ExtensionPoint interface that represents a priority that the class searcher can use to sort on. 

Thoughts?

**Comment by astrelsky** [comment]
> We are thinking:
> 
> 1. Sort the classpath that both the class loader and class searcher use according to some priority scheme (something like group frameworks, then features, then processors, then extensions, and subsort each group by module name)
> 2. Add an annotation to the ExtensionPoint interface that represents a priority that the class searcher can use to sort on.
> 
> Thoughts?

Seems fairly straightforward and almost identical to how the AbstractLibrarySupportLoader works. My only question is if you were to extend an extension point to slight modify its behavior for a specific processor, would you be able to get the priority of the of the super class? I'm not familiar with annotations in Java yet, I only know what they are from decorators in python.

For example to squash bad relocations without having to rewrite the ElfProgramBuilder I just did the following:

```
public class DummyLoader extends ElfLoader {

    ...

    @Override
    public int getTierPriority() {
        return super.getTierPriority()-1;
    }

}
```

**Comment by ryanmkurtz** [comment]
Sorry, I should have provided more context.  Module names are things like Base, Generic, Python, etc.  Things in the Features, Frameworks, Processors, and Extensions directory.  The module name is the directory name, and Ghidra does not support duplicate module names.  

The purpose of 1) was simply to provide a deterministic classpath ordering from run to run via sorting.  This would solve your problem of your extension point sometimes running and sometimes not (it's also something we should have been doing from the start anyway).

The purpose of 2) is so that you don't have to worry about how to name your class to get it chosen by the class searcher.  You would simply provide a higher priority on your extension point via a java annotation.

**Comment by astrelsky** [comment]
> Sorry, I should have provided more context. Module names are things like Base, Generic, Python, etc. Things in the Features, Frameworks, Processors, and Extensions directory. The module name is the directory name, and Ghidra does not support duplicate module names.
> 
> The purpose of 1) was simply to provide a deterministic classpath ordering from run to run via sorting. This would solve your problem of your extension point sometimes running and sometimes not (it's also something we should have been doing from the start anyway).
> 
> The purpose of 2) is so that you don't have to worry about how to name your class to get it chosen by the class searcher. You would simply provide a higher priority on your extension point via a java annotation.

I see, this make much more sense. I do have one question regarding number 2 though. To be clear, I don't know much about annotations in Java. I only know what they are from experience with decorators in Python.

Will the priority provided by the annotation be hard-coded or will it work in a similar fashion to the analysis priorities? If they are hard coded to a number, would it be possible to fetch the priority of a class being extended? Being able to do so would ensure you still have priority if the priority of the superclass is ever changed.

For example this is what I've done with an AbstractLibrarySupportLoader which is extending ElfLoader:

```
public class DummyLoader extends ElfLoader {
    ...

    @Override
    public int getTierPriority() {
        return super.getTierPriority()-1;
    }

}
```

**Comment by ryanmkurtz** [comment]
The annotations will likely be hardcoded integers like the the Loader's tier priority, and you will be able to get any extension point's priority to do things like your example.

**Comment by astrelsky** [comment]
> The annotations will likely be hardcoded integers like the the Loader's tier priority, and you will be able to get any extension point's priority to do things like your example.

Sounds like a plan then. I'm assuming for ease of integration a default value will be provided to the ExtensionPoint Interface?

**Comment by ryanmkurtz** [comment]
Yes, I don't plan on adding any priorities to things that ship with Ghidra.  They would all just get the default value.

**Comment by ryanmkurtz** [comment]
Fixed by 8f001526010c4a8399d864c6dace5358b639ce57.

You can now add something like

`@ExtensionPointProperties(priority = 2)`

to your extension point class.  Higher priority #'s will get used first (the default for all built-in Ghidra extension points is 1).
